azure-iot-device/azure/iot/device/common/pipeline/pipeline_ops_base.py

```diff
@@ -162,3 +162,82 @@ def __init__(self, sas_token, callback=None):
         """
         super(SetSasToken, self).__init__(callback=callback)
         self.sas_token = sas_token
+
+
+class SendIotRequestAndWaitForResponse(PipelineOperation):
+    """
+    A PipelineOperation object which wraps the common operation of sending a request to iothub with a request_id ($rid)
+    value and waiting for a response with the same $rid value.  This convention is used by both Twin and Provisioning
+    features.
+
+    Even though this is an base operation, it will most likely be generated and also handled by more specifics stages
+    (such as IotHub or Mqtt stages).
+
+    The type of the request payload and the response payload is undefined at this level.  The type of the payload is defined
+    based on the type of request that is being executed.  If types need to be converted, that is the responsibility of
+    the stage which creates this operation, and also the stage which executes on the operation.
+
+    :ivar status_code: The status code returned by the response.  Any value under 300 is considered success.
+    :type status_code: int
+    :ivar response_body: The body of the response.
+    :type response_payload: Undefined
+    """
+
+    def __init__(self, request_type, verb, resource, request_body, callback=None):
+        """
+        Initializer for SendIotRequestAndWaitForResponse objects
+
+        :param str request_type: The type of request.  This is a string which is used by transport-specific stages to
+          generate the actual request.  For example, if request_type is "twin", then the iothub_mqtt stage will convert
+          the request into an MQTT publish with topic that begins with $iothub/twin
+        :param str verb: The verb for the request, such as "POST", "GET", etc.
```

Comment:
> :param str verb: The verb for the request, such as "POST", "GET", etc. [](start = 7, length = 71)

1) I think the correct term for these are HTTP methods, not verbs

2) What does `POST` or `GET` even mean in the context of MQTT?

Reply:
I've seen the term "method" used in other codebases.  I'll change it.

These method names mean nothing in the concept of MQTT, but they do have meaning in the context of iothub, where the convention is to use MQTT and AMQP following a pattern that is like "REST-over-MQTT" or "REST-over-AMQP".  in this way, the iothub protocols are designed for the least-common-denominator HTTP.  They don't really take advantage of MQTT and AMQP as much as they could.  This is why provioning, for instance, needs to poll the service for status instead of relying on a service-push.

---
In reply to: [288341305](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288341305) [](ancestors = 288341305)

Reply:
It's unfortunate that "method" also isn't a great name, since it can be confused for the method of an object... but I think the docs make it clear, so I'm okay with it

---
In reply to: [288663168](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288663168) [](ancestors = 288663168,288341305)